### PR TITLE
WT-7159 Always write on-disk update as a full update to history store

### PR DESCRIPTION
**Pull request by kommiharibabu**
The updates written to the data store are always full updates, but
when these updates are moved into the history store by eviction due
to further updates on the key, make sure that on-disk update is always
written as full update even in the update list it is a modify, because
the on-disk update will be the base update for all the history store
updates that are previous to the on-disk update.

Due to concurrent operation of checkpoint and eviction, it is possible
that history store may have more recent versions of a key than the
on-disk version. Without a proper base value in the history store,
it can lead to wrong value being restored by the RTS.

### CODE REVIEW

**Comment by kommiharibabu** [comment]
@bvpvamsikrishna @quchenhao @agorrod I wasn't able to generate the corruption scenario by applying the same timestamped updates to the base value, still, I kept the test as part of the PR as it adds value to the testing. I will keep trying to generate a corruption scenario and update the test later.

File: src/history/hs_rec.c [hunk 1]
```diff
@@ -686,7 +686,8 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
              * inserting the first update.
              */
             nentries = MAX_REVERSE_MODIFY_NUM;
-            if (upd->type == WT_UPDATE_MODIFY && enable_reverse_modify &&
+            if (!F_ISSET(upd, WT_UPDATE_DS) && upd->type == WT_UPDATE_MODIFY &&
```
Comment (hbokhari) [inline src/history/hs_rec.c:689]: I believe a comment around why we alway want to insert a previous datastore value as a full update will be helpful for future reference.

Comment (kommiharibabu) [reply src/history/hs_rec.c:689]: Sure will update the comment to explain the reason behind.

**Review by quchenhao — APPROVED** [review]
LGTM if the comment is addressed.

Not sure we should keep the previous timestamp check in rollback to stable. At least we should modify its comment to record why we still keep that check.

File: src/history/hs_rec.c (hunk 1 above)
Comment (kommiharibabu) [reply src/history/hs_rec.c:703]: Added a comment.

**Comment by kommiharibabu** [comment]
Added a comment in a rollback to stable to explain why it doesn't cause any problem even if we process the most recent updates than on-disk, but still, it is better to ignore them as it unnecessarily adds to rollback to stable time.

**Comment by kommiharibabu** [comment]
The stress tests and patch builds are passed.

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

File: src/history/hs_rec.c [hunk 2]
```diff
@@ -690,10 +690,18 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
 
             /*
              * Calculate reverse modify and clear the history store records with timestamps when
-             * inserting the first update.
+             * inserting the first update. Always write on-disk data store updates to the history
+             * store as a full update, because the on-disk update will be the base update for all
```
Comment (hbokhari) [inline src/history/hs_rec.c:694]: update, because -> update because

File: src/history/hs_rec.c [hunk 3]
```diff
@@ -690,10 +690,18 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
+             * the updates are previous to the on-disk update.
```
Comment (hbokhari) [inline src/history/hs_rec.c:695]: updates are previous to the on-disk update. -> older than the previous on-disk update.

File: src/txn/txn_rollback_to_stable.c [hunk 4]
```diff
@@ -254,10 +254,15 @@ __rollback_row_ondisk_fixup_key(WT_SESSION_IMPL *session, WT_PAGE *page, WT_ROW
 
         /*
          * Do not include history store updates greater than on-disk data store version to construct
-         * a full update to restore. Comparing with timestamps here has no problem unlike in search
-         * flow where the timestamps may be reset during reconciliation. RTS detects an on-disk
-         * update is unstable based on the written proper timestamp, so comparing against it with
-         * history store shouldn't have any problem.
+         * a full update to restore. Even if we include the most recent updates than the on-disk
```
Comment (hbokhari) [inline src/txn/txn_rollback_to_stable.c:257]: Even if we include the most recent updates than the on-disk version shouldn't be problem -> Including more recent updates than the on-disk version shouldn't be a problem

**Review by hbokhari — APPROVED** [review]
LGTM with minor fixes in the comments.

File: src/txn/txn_rollback_to_stable.c (hunk 4 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed.

File: src/history/hs_rec.c (hunk 3 above)
Comment (kommiharibabu) [reply src/history/hs_rec.c]: Fixed.

File: src/history/hs_rec.c (hunk 2 above)
Comment (kommiharibabu) [reply src/history/hs_rec.c]: Fixed.

**Comment by kommiharibabu** [comment]
The test failure is not related to the PR changes.

**Pull request merged**